File: Excel/554 Challenge.R
library(tidyverse)
library(readxl)
library(parallel)

path = "Excel/554 Kaprekar Numbers.xlsx"
test = read_excel(path, range = "A1:A51")

check_kaprekar_fast = function(n) {
  nsqr = n^2
  digits = floor(log10(nsqr)) + 1
  for (split_pos in 1:(digits - 1)) {
    right_part = nsqr %% 10^split_pos
    left_part = nsqr %/% 10^split_pos
    if (right_part > 0 && left_part + right_part == n) {
      return(TRUE)
    }
  }
  return(FALSE)
}

parallel_kaprekar_check = function(n_values) {
  num_cores = detectCores() - 1
  cl = makeCluster(num_cores)
  clusterExport(cl, "check_kaprekar_fast")
  result = parLapply(cl, n_values, check_kaprekar_fast)
  stopCluster(cl)
  return(unlist(result))
}

n_values = 4:1000000
kaprekar_flags = parallel_kaprekar_check(n_values)

df = data.frame(n = n_values, is_kaprekar = kaprekar_flags) %>%
  filter(is_kaprekar) %>%
  head(50) %>%
  select(n)

all.equal(df, test, check.attributes = FALSE) # TRUE

File: Excel/500-599/554/554 Challenge.py
import pandas as pd
import math

path = "554 Kaprekar Numbers.xlsx"
test = pd.read_excel(path, usecols="A", nrows=51)

def check_kaprekar_fast(n):
    nsqr = n ** 2
    digits = math.floor(math.log10(nsqr)) + 1
    for split_pos in range(1, digits):
        right_part = nsqr % (10 ** split_pos)
        left_part = nsqr // (10 ** split_pos)
        if right_part > 0 and left_part + right_part == n:
            return True
    return False

n_values = range(4, 1000001)
kaprekar_flags = [check_kaprekar_fast(n) for n in n_values]

df = pd.DataFrame({'n': n_values, 'is_kaprekar': kaprekar_flags})
df = df[df['is_kaprekar']].head(50).loc[:, ['n']].reset_index(drop=True)

print(df["n"].equals(test["Answer Expected"])) # True